File: eventual-consistency-support-core/src/main/java/org/xyattic/eventual/consistency/support/core/aop/support/AnnotationClassOrMethodBeanFactoryPointcutAdvisor.java
package org.xyattic.eventual.consistency.support.core.aop.support;

import org.springframework.aop.Pointcut;
import org.springframework.aop.support.AbstractBeanFactoryPointcutAdvisor;
import org.springframework.aop.support.ComposablePointcut;

import java.lang.annotation.Annotation;
import java.util.Arrays;

/**
 * @author wangxing
 * @create 2020/11/13
 */
public class AnnotationClassOrMethodBeanFactoryPointcutAdvisor extends AbstractBeanFactoryPointcutAdvisor {

    private ComposablePointcut pointcut;

    @SafeVarargs
    public AnnotationClassOrMethodBeanFactoryPointcutAdvisor(Class<? extends Annotation>... annotationTypes) {
        Arrays.stream(annotationTypes)
                .map(AnnotationClassOrMethodPointcut::new)
                .forEach(classOrMethodPointcut -> {
                    if (pointcut == null) {
                        pointcut = new ComposablePointcut(classOrMethodPointcut);
                    } else {
                        pointcut.union(classOrMethodPointcut);
                    }
                });
        if (pointcut == null) {
            pointcut = new ComposablePointcut();
        }
    }

    public AnnotationClassOrMethodBeanFactoryPointcutAdvisor(AnnotationMethodInterceptor<?> annotationMethodInterceptor) {
        this(annotationMethodInterceptor.getAnnotationType());
        setAdvice(annotationMethodInterceptor);
        setOrder(annotationMethodInterceptor.getOrder());
    }

    @Override
    public Pointcut getPointcut() {
        return pointcut;
    }
}

File: eventual-consistency-support-core/src/main/java/org/xyattic/eventual/consistency/support/core/aop/support/AnnotationClassOrMethodBeanFactoryPointcutAdvisor.kt
package org.xyattic.eventual.consistency.support.core.aop.support

import org.springframework.aop.Pointcut
import org.springframework.aop.support.AbstractBeanFactoryPointcutAdvisor
import org.springframework.aop.support.ComposablePointcut
import java.util.*

/**
 * @author wangxing
 * @create 2020/11/13
 */
open class AnnotationClassOrMethodBeanFactoryPointcutAdvisor
@SafeVarargs constructor(vararg annotationTypes: Class<out Annotation>) :
        AbstractBeanFactoryPointcutAdvisor() {

    private var pointcut: ComposablePointcut

    constructor(annotationMethodInterceptor: AnnotationMethodInterceptor<*>) :
            this(annotationMethodInterceptor.annotationType) {
        advice = annotationMethodInterceptor
        order = annotationMethodInterceptor.order
    }

    override fun getPointcut(): Pointcut {
        return pointcut
    }

    init {
        var pointcut: ComposablePointcut? = null
        Arrays.stream(annotationTypes)
                .map { annotationType: Class<out Annotation> -> AnnotationClassOrMethodPointcut(annotationType) }
                .forEach { classOrMethodPointcut: AnnotationClassOrMethodPointcut ->
                    pointcut = pointcut?.union(classOrMethodPointcut)
                            ?: ComposablePointcut(classOrMethodPointcut)
                }
        this.pointcut = pointcut ?: ComposablePointcut()
    }

}